File: backend/app/mathematics/generator.py
from __future__ import annotations

import logging
import uuid
from collections.abc import Generator

from app.mathematics.rng import create_rng
from app.mathematics.tiers import TIER_CONFIGS, TierConfig
from app.mathematics.types import Operation, Problem, ProblemSet

logger = logging.getLogger(__name__)

_MAX_RETRIES = 10


def _random_int(rng: Generator[float, None, None], lo: int, hi: int) -> int:
    return int(next(rng) * (hi - lo + 1)) + lo


def _pick_operation(config: TierConfig, rng: Generator[float, None, None]) -> Operation:
    ops = config.operations
    return ops[int(next(rng) * len(ops))]


def _pick_operands(
    operation: Operation, config: TierConfig, rng: Generator[float, None, None]
) -> tuple[int, int]:
    lo, hi = config.min_operand, config.max_operand
    if operation == Operation.division:
        b = _random_int(rng, 2, hi)
        max_multiplier = max(1, hi // b)
        a = b * _random_int(rng, 1, max_multiplier)
        return a, b
    if operation == Operation.subtraction:
        a = _random_int(rng, lo, hi)
        b = _random_int(rng, lo, a)
        return a, b
    return _random_int(rng, lo, hi), _random_int(rng, lo, hi)


def _compute(operation: Operation, a: int, b: int) -> int:
    if operation == Operation.addition:
        return a + b
    if operation == Operation.subtraction:
        return a - b
    if operation == Operation.multiplication:
        return a * b
    return a // b


def _is_duplicate(candidate: Problem, last: Problem | None) -> bool:
    if last is None:
        return False
    return (
        candidate.operation == last.operation
        and candidate.operand_a == last.operand_a
        and candidate.operand_b == last.operand_b
    )


def generate_problem_set(
    tier: int,
    seed: int,
    count: int,
    custom_tier_config: TierConfig | None = None,
) -> ProblemSet:
    if tier == 6:
        config = custom_tier_config or TierConfig(
            6,
            TIER_CONFIGS[5].operations,
            TIER_CONFIGS[5].min_operand,
            TIER_CONFIGS[5].max_operand,
        )
    else:
        config = TIER_CONFIGS[tier]

    rng = create_rng(seed)
    problems: list[Problem] = []
    last: Problem | None = None

    while len(problems) < count:
        problem: Problem | None = None
        for attempt in range(_MAX_RETRIES + 1):
            operation = _pick_operation(config, rng)
            a, b = _pick_operands(operation, config, rng)
            answer = _compute(operation, a, b)
            candidate = Problem(
                id=uuid.uuid4(),
                operation=operation,
                operand_a=a,
                operand_b=b,
                answer=answer,
                tier=tier,
                seed=seed,
            )
            if not _is_duplicate(candidate, last) or attempt == _MAX_RETRIES:
                if attempt == _MAX_RETRIES and _is_duplicate(candidate, last):
                    logger.warning(
                        "Retry limit reached — accepting duplicate problem",
                        extra={"context": {"tier": tier, "seed": seed, "slot": len(problems)}},
                    )
                problem = candidate
                break
        problems.append(problem)  # type: ignore[arg-type]
        last = problem

    return ProblemSet(seed=seed, tier=tier, count=count, problems=tuple(problems))

File: frontend/src/engine/math/generator.ts
import { createRng } from './rng';
import { TIER_CONFIGS } from './tiers';
import type { Operation, Problem, ProblemSet, Tier, TierConfig } from './types';

const MAX_RETRIES = 10;

function randomInt(rng: () => number, min: number, max: number): number {
  return Math.floor(rng() * (max - min + 1)) + min;
}

function pickOperation(config: TierConfig, rng: () => number): Operation {
  const ops = config.operations;
  return ops[Math.floor(rng() * ops.length)];
}

function pickOperands(
  operation: Operation,
  config: TierConfig,
  rng: () => number,
): [number, number] {
  const { minOperand, maxOperand } = config;
  if (operation === 'division') {
    const b = randomInt(rng, 2, maxOperand);
    const maxMultiplier = Math.floor(maxOperand / b);
    const a = b * randomInt(rng, 1, Math.max(1, maxMultiplier));
    return [a, b];
  }
  if (operation === 'subtraction') {
    const a = randomInt(rng, minOperand, maxOperand);
    const b = randomInt(rng, minOperand, a);
    return [a, b];
  }
  return [randomInt(rng, minOperand, maxOperand), randomInt(rng, minOperand, maxOperand)];
}

function compute(operation: Operation, a: number, b: number): number {
  switch (operation) {
    case 'addition':
      return a + b;
    case 'subtraction':
      return a - b;
    case 'multiplication':
      return a * b;
    case 'division':
      return a / b;
  }
}

function isDuplicate(candidate: Problem, last: Problem | null): boolean {
  if (last === null) return false;
  return (
    candidate.operation === last.operation &&
    candidate.operand_a === last.operand_a &&
    candidate.operand_b === last.operand_b
  );
}

export function generateProblemSet(
  tier: Tier,
  seed: number,
  count: number,
  customTierConfig?: TierConfig,
): ProblemSet {
  const config: TierConfig =
    tier === 6
      ? customTierConfig ?? { ...TIER_CONFIGS[5], tier: 6 }
      : TIER_CONFIGS[tier as Exclude<Tier, 6>];

  const rng = createRng(seed);
  const problems: Problem[] = [];
  let last: Problem | null = null;

  while (problems.length < count) {
    let problem: Problem | null = null;
    for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
      const operation = pickOperation(config, rng);
      const [a, b] = pickOperands(operation, config, rng);
      const answer = compute(operation, a, b);
      const candidate: Problem = {
        id: crypto.randomUUID(),
        operation,
        operand_a: a,
        operand_b: b,
        answer,
        tier,
        seed,
      };
      if (!isDuplicate(candidate, last) || attempt === MAX_RETRIES) {
        problem = candidate;
        break;
      }
    }
    problems.push(problem!);
    last = problem;
  }

  return { seed, tier, count, problems };
}
